### PR TITLE
8280510: AArch64: Vectorize operations with loop induction variable

### PR DESCRIPTION
**Pull request by pfustc**
AArch64 has SVE instruction of populating incrementing indices into an
SVE vector register. With this we can vectorize some operations in loop
with the induction variable operand, such as below.
```
  for (int i = 0; i < count; i++) {
    b[i] = a[i] * i;
  }
```
This patch enables the vectorization of operations with loop induction
variable by extending current scope of C2 superword vectorizable packs.
Before this patch, any scalar input node in a vectorizable pack must be
an out-of-loop invariant. This patch takes the induction variable input
as consideration. It allows the input to be the iv phi node or phi plus
its index offset, and creates a `PopulateIndexNode` to generate a vector
filled with incrementing indices. On AArch64 SVE, final generated code
for above loop expression is like below.
```
  add     x12, x16, x10
  add     x12, x12, #0x10
  ld1w    {z16.s}, p7/z, [x12]
  index   z17.s, w1, #1
  mul     z17.s, p7/m, z17.s, z16.s
  add     x10, x17, x10
  add     x10, x10, #0x10
  st1w    {z17.s}, p7, [x10]
```
As there is no populating index instruction on AArch64 NEON or other
platforms like x86, a function named `is_populate_index_supported()` is
created in the VectorNode class for the backend support check.

Jtreg hotspot::hotspot_all_no_apps, jdk::tier1~3 and langtools::tier1
are tested and no issue is found. Hotspot jtreg has existing tests in
`compiler/c2/cr7192963/Test*Vect.java` covering this kind of use cases so
no new jtreg is created within this patch. A new JMH is created in this
patch and tested on a 512-bit SVE machine. Below test result shows the
performance can be significantly improved in some cases.
```
  Benchmark                       Performance
  IndexVector.exprWithIndex1            ~7.7x
  IndexVector.exprWithIndex2           ~13.3x
  IndexVector.indexArrayFill            ~5.7x
```

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8280510](https://bugs.openjdk.java.net/browse/JDK-8280510): AArch64: Vectorize operations with loop induction variable

### Reviewers
 * [Andrew Dinn](https://openjdk.java.net/census#adinn) (@adinn - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7491/head:pull/7491` \
`$ git checkout pull/7491`

Update a local copy of the PR: \
`$ git checkout pull/7491` \
`$ git pull https://git.openjdk.java.net/jdk pull/7491/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7491`

View PR using the GUI difftool: \
`$ git pr show -t 7491`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7491.diff">https://git.openjdk.java.net/jdk/pull/7491.diff</a>

</details>

### CODE REVIEW

**Comment by pfustc** [comment]
Can anyone help review this? This optimization is currently for AArch64 but changes are mainly in C2 mid-end.

File: src/hotspot/share/opto/superword.cpp [hunk 1]
```diff
@@ -1316,7 +1316,16 @@ bool SuperWord::have_similar_inputs(Node* s1, Node* s2) {
...
-      if (s1->in(i)->Opcode() != s2->in(i)->Opcode()) return false;
+      Node* s1_in = s1->in(i);
+      Node* s2_in = s2->in(i);
+      if (s1_in->is_Phi() && s2_in->is_Add() && s2_in->in(1) == s1_in) {
+        // Special handling for expressions with loop iv, like "b[i] = a[i] * i".
+        // In this case, one node has an input from the tripcount iv and another
+        // node has an input from iv plus an offset.
+        if (!s1_in->as_Phi()->is_tripcount(T_INT)) return false;
+      } else {
+        if (s1_in->Opcode() != s1_in->Opcode()) return false;
```
Comment (adinn) [inline src/hotspot/share/opto/superword.cpp:1327]: This last line ought to be `s1_in->Opcode() != s2_in->Opcode` but you have replaced `s2` with `s1`.

**Review by adinn — COMMENTED** [review]
The code looks ok to me modulo the cut-and-paste error I highlighted.

Have you any way of ensuring that the generated code is correct (other than the fact that it runs quicker).

Also have you checked this does not invalidate any other SVE tests. I would expect some failures thanks to the cut-and-paste error.

**Comment by adinn** [comment]
It would be a good idea if someone who knows the superword code better than me could comment on this extension to the range of cases covered by superword transformation. This looks like a very useful extension to the range of cases that can be optimized using superword analysis. However, I am not familiar enough with the code to know if this is the best way to proceed or if some other alternative has been considered and might provide a better alternative, especially wrt architectures other than aarch64.

**Comment by pfustc** [comment]
> The code looks ok to me modulo the cut-and-paste error I highlighted.

Thanks for your careful review. The cut-and-paste error has been fixed in my updated patch.

> Have you any way of ensuring that the generated code is correct (other than the fact that it runs quicker).
>
> Also have you checked this does not invalidate any other SVE tests. I would expect some failures thanks to the cut-and-paste error.

I've run almost full jtreg on an SVE featured machine and no issue is found. I noticed that existing hotspot jtregs cases in `compiler/c2/cr7192963/Test*Vect.java` have this kind of loop patterns. (Unfortunately, there is no related failure found even with that cut-and-paste error.)

> It would be a good idea if someone who knows the superword code better than me could comment on this extension to the range of cases covered by superword transformation. This looks like a very useful extension to the range of cases that can be optimized using superword analysis. However, I am not familiar enough with the code to know if this is the best way to proceed or if some other alternative has been considered and might provide a better alternative, especially wrt architectures other than aarch64.

I agree. I think people who know superword well are mostly in Oracle so I'd like to see if anyone in Oracle could help look at this. I've considered enabling this for x86. But as I'm not familiar with x86 instruction set enough, I haven't found any instruction(s) to implement such index vector. I would be very glad if someone from Intel can give me some suggestions.

**Comment by pfustc** [comment]
@vnkozlov @TobiHartmann Can you also help look at this?

**Comment by TobiHartmann** [comment]
Please resolve the merge conflicts.

**Comment by pfustc** [comment]
> Please resolve the merge conflicts.

Done. After merge, this is also tested with recent new jtreg cases in `compiler/vectorization/runner/ArrayIndexFillTest.java`

**Review by TobiHartmann — COMMENTED** [review]
I executed some testing, all passed.

Do we need to add the new node to `MatchRule::is_expensive()`?

Do we need to add a declaration to `vmStructs.cpp`?

**Comment by pfustc** [comment]
Hi @TobiHartmann ,

> Do we need to add the new node to MatchRule::is_expensive()?
> 
> Do we need to add a declaration to vmStructs.cpp?

I've fixed these in my latest commit. In that commit, I also align matching rule code with other rules in AArch64 ad file.

**Review by adinn — APPROVED** [review]
Still looks good to me

**Review by TobiHartmann — APPROVED** [review]
Looks good to me too. Have you considered adding a IR verification test?

**Comment by pfustc** [comment]
Hi @TobiHartmann ,

> Looks good to me too. Have you considered adding a IR verification test?

Thanks for your review. And good question!

You might have remembered that we contributed a vectorization test framework together with our post loop fix (https://github.com/openjdk/jdk/pull/6828). Currently we are enabling IR tests in that framework. So far we have made some progress. So IR tests will be added for all vectorizable loops there (in `hotspot:compiler/vectorization/runner/`) in the near future.

**Pull request closed without merging**